src/test/prevector_tests.cpp

```diff
@@ -205,13 +202,11 @@ class prevector_tester {
     }
 
     ~prevector_tester() {
-        BOOST_CHECK_MESSAGE(passed, "insecure_rand: " + rand_seed.ToString());
+        BOOST_CHECK_MESSAGE(passed, "The random seed was logged at the start of the test");
```

Comment:
In commit "test: Correct the random seed log on a prevector test failure" (faa2608e6c3e1763d0991443c3ce6d9bb1466da8)

I think it would be nice to just fix the bug and print the seed when the test fails rather than requiring looking up the seed in logs. It seems like we could easily support this by adding `return seed` as the last line of `SeedRandomForTest` and this would be helpful for debugging tests that only fail with certain seeds, because they could be run without always enabling logging.

Reply:
re: https://github.com/bitcoin/bitcoin/pull/30571#discussion_r1725675521

That all sounds good but I think I'm proposing a smaller not bigger change in behavior by just fixing the bug instead of trying to do something grander. The change based on master is:

```diff
--- a/src/test/prevector_tests.cpp
+++ b/src/test/prevector_tests.cpp
@@ -209,9 +209,8 @@ public:
     }
 
     prevector_tester() {
-        SeedRandomForTest();
-        rand_seed = InsecureRand256();
-        rand_cache.Reseed(rand_seed);
+        rand_seed = SeedRandomForTest();
+        rand_cache.Reseed(InsecureRand256());
     }
 };
 
--- a/src/test/util/random.cpp
+++ b/src/test/util/random.cpp
@@ -15,7 +15,7 @@ FastRandomContext g_insecure_rand_ctx;
 
 extern void MakeRandDeterministicDANGEROUS(const uint256& seed) noexcept;
 
-void SeedRandomForTest(SeedRand seedtype)
+uint256 SeedRandomForTest(SeedRand seedtype)
 {
     static const std::string RANDOM_CTX_SEED{"RANDOM_CTX_SEED"};
 
@@ -35,4 +35,5 @@ void SeedRandomForTest(SeedRand seedtype)
     LogPrintf("%s: Setting random seed for current tests to %s=%s\n", __func__, RANDOM_CTX_SEED, seed.GetHex());
     MakeRandDeterministicDANGEROUS(seed);
     g_insecure_rand_ctx.Reseed(GetRandHash());
+    return seed;
 }
--- a/src/test/util/random.h
+++ b/src/test/util/random.h
@@ -25,7 +25,7 @@ enum class SeedRand {
 };
 
 /** Seed the RNG for testing. This affects all randomness, except GetStrongRandBytes(). */
-void SeedRandomForTest(SeedRand seed = SeedRand::SEED);
+uint256 SeedRandomForTest(SeedRand seed = SeedRand::SEED);
 
 static inline uint32_t InsecureRand32()
 {

```

This would change the test less than the current PR, provide nicer behavior, and help out with future changes that make the seed more accessible, by returning the value from the function which set it.

Reply:
re: https://github.com/bitcoin/bitcoin/pull/30571#discussion_r1725758066

> Starting with commit https://github.com/bitcoin/bitcoin/commit/fae43a97ca947cd0802392e9bb86d9d0572c0fba, the seed is static const per process

Wow, I didn't notice `ctx_seed` was statically caching a random value before this so I've been misunderstanding what the `SeedRandomForTest` function has been doing the whole time. I think this could be made clearer by:

- Renaming `SeedRand::SEED` to `SeedRand::FIXED_SEED` so it is clearer `SeedRandomForTest` sets a fixed value, not one that can change between calls.
- Dropping the `SeedRandomForTest` default argument value so if a test resets to the fixed seed the calls look like `SeedRandomForTest(SeedRand::FIXED_SEED)` which tells you a fixed value is being set, as opposed to `SeedRandomForTest()` which makes it look like it could be setting a new random seed.
- Adding comments that better document what the seed functions and enum values do.

Would suggest:

```diff
--- a/src/test/util/random.h
+++ b/src/test/util/random.h
@@ -13,10 +13,13 @@
 
 enum class SeedRand {
     ZEROS, //!< Seed with a compile time constant of zeros
-    SEED,  //!< Use (and report) random seed from environment, or a (truly) random one.
+    FIXED_SEED, //!< Seed with a fixed value that never changes over the
+                //!< lifetime of this process. The seed is read from the
+                //!< RANDOM_CTX_SEED environment variable if set, otherwise
+                //!< generated randomly once, saved, and reused.
 };
 
-/** Seed the internal global RNG state for testing. This affects all randomness, except GetStrongRandBytes(). */
+/** Seed the global RNG state for testing and log the seed value. This affects all randomness, except GetStrongRandBytes(). */
 void SeedRandomStateForTest(SeedRand seed);
 
 template <RandomNumberGenerator Rng>
--- a/src/test/util/setup_common.h
+++ b/src/test/util/setup_common.h
@@ -67,8 +67,8 @@ struct BasicTestingSetup {
     node::NodeContext m_node; // keep as first member to be destructed last
 
     FastRandomContext m_rng;
-    /** Seed the internal global RNG state and m_rng for testing. This affects all randomness, except GetStrongRandBytes(). */
-    void SeedRandomForTest(SeedRand seed = SeedRand::SEED)
+    /** Seed the global RNG state and m_rng for testing and log the seed value. This affects all randomness, except GetStrongRandBytes(). */
+    void SeedRandomForTest(SeedRand seed)
     {
         SeedRandomStateForTest(seed);
         m_rng.Reseed(GetRandHash());

```